### PR TITLE
[Merged by Bors] - feat: port Algebra.Ring.Pi

### PR DESCRIPTION
**Pull request by joneugster**

### CODE REVIEW

**Comment by eric-wieser** [comment]
Note there is a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Algebra.2ERing.2EPi/near/317155120) about this file, which suggests we should eliminate the use of, rather than wait for, these tactics.

File: Mathlib/Algebra/Ring/Pi.lean [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+instance nonUnitalRing [∀ i, NonUnitalRing <| f i] : NonUnitalRing (∀ i : I, f i) :=
+  { Pi.nonUnitalNonAssocRing, Pi.nonUnitalSemiring with }
+#align pi.non_unital_ring Pi.nonUnitalRing
+
+instance nonAssocRing [∀ i, NonAssocRing <| f i] : NonAssocRing (∀ i : I, f i) :=
+  { Pi.nonUnitalNonAssocRing, Pi.nonAssocSemiring with}
+#align pi.non_assoc_ring Pi.nonAssocRing
+
+instance ring [∀ i, Ring <| f i] : Ring (∀ i : I, f i) :=
+  { Pi.semiring, Pi.addCommGroup, Pi.addGroupWithOne with}
```
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Pi.lean:83]: Untested,
```suggestion
  { Pi.semiring, Pi.addCommGroupWithOne with}
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#align pi.ring Pi.ring
+
+instance nonUnitalCommRing [∀ i, NonUnitalCommRing <| f i] : NonUnitalCommRing (∀ i : I, f i) :=
+  { Pi.nonUnitalRing, Pi.commSemigroup with }
+#align pi.non_unital_comm_ring Pi.nonUnitalCommRing
+
+instance commRing [∀ i, CommRing <| f i] : CommRing (∀ i : I, f i) :=
+  { Pi.ring, Pi.commMonoid with }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Pi.lean:91]: My leaning would be
```suggestion
  { Pi.ring, Pi.commSemiring with }
```
and generally use the largest possible base class, but it's not that important.

Do you know if this type of thing already discussed on Zulip?

Comment (joneugster) [reply Mathlib/Algebra/Ring/Pi.lean:91]: I'm not aware of such a discussion, but makes sense, I'll change that.

File: Mathlib/Algebra/Ring/Pi.lean (hunk 1 above)
Comment (joneugster) [reply Mathlib/Algebra/Ring/Pi.lean:83]: `Algebra.Group.Pi` doesn't import `AddCommGroupWithOne`  and therefore is also not defining `Pi.addCommGroupWithOne`.

**Review by mcdoll — APPROVED** [review]
LGTM, but I am biased in so far that I really want this merged.

maintainer merge

File: Mathlib/Algebra/Ring/Pi.lean [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+end NonUnitalRingHom
+
+section RingHom
+
+universe u v
+
+variable {I : Type u}
+
+/-- Evaluation of functions into an indexed collection of rings at a point is a ring
+homomorphism. This is `function.eval` as a `ring_hom`. -/
```
Comment (mcdoll) [inline Mathlib/Algebra/Ring/Pi.lean:164]:
```suggestion
homomorphism. This is `function.eval` as a `RingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+@[simps]
+def Pi.evalRingHom (f : I → Type v) [∀ i, NonAssocSemiring (f i)] (i : I) : (∀ i, f i) →+* f i :=
+  { Pi.evalMonoidHom f i, Pi.evalAddMonoidHom f i with }
+#align pi.eval_ring_hom Pi.evalRingHom
+
+/-- `function.const` as a `ring_hom`. -/
```
Comment (mcdoll) [inline Mathlib/Algebra/Ring/Pi.lean:170]:
```suggestion
/-- `function.const` as a `RingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+end Pi
+
+section NonUnitalRingHom
+
+universe u v
+
+variable {I : Type u}
+
+/-- Evaluation of functions into an indexed collection of non-unital rings at a point is a
+non-unital ring homomorphism. This is `function.eval` as a `nonUnitalRingHom`. -/
```
Comment (mcdoll) [inline Mathlib/Algebra/Ring/Pi.lean:134]:
```suggestion
non-unital ring homomorphism. This is `function.eval` as a `NonUnitalRingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+@[simps]
+def Pi.evalNonUnitalRingHom (f : I → Type v) [∀ i, NonUnitalNonAssocSemiring (f i)] (i : I) :
+    (∀ i, f i) →ₙ+* f i :=
+  { Pi.evalMulHom f i, Pi.evalAddMonoidHom f i with }
+#align pi.eval_non_unital_ring_hom Pi.evalNonUnitalRingHom
+
+/-- `function.const` as a `nonUnitalRingHom`. -/
```
Comment (mcdoll) [inline Mathlib/Algebra/Ring/Pi.lean:141]:
```suggestion
/-- `function.const` as a `NonUnitalRingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+end NonUnitalRingHom
+
+section RingHom
+
+universe u v
+
+variable {I : Type u}
+
+/-- Evaluation of functions into an indexed collection of rings at a point is a ring
+homomorphism. This is `function.eval` as a `RingHom`. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Pi.lean:164]:
```suggestion
homomorphism. This is `Function.eval` as a `RingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+@[simps]
+def Pi.evalRingHom (f : I → Type v) [∀ i, NonAssocSemiring (f i)] (i : I) : (∀ i, f i) →+* f i :=
+  { Pi.evalMonoidHom f i, Pi.evalAddMonoidHom f i with }
+#align pi.eval_ring_hom Pi.evalRingHom
+
+/-- `function.const` as a `RingHom`. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Pi.lean:170]:
```suggestion
/-- `Function.const` as a `RingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+/-- Evaluation of functions into an indexed collection of non-unital rings at a point is a
+non-unital ring homomorphism. This is `function.eval` as a `NonUnitalRingHom`. -/
+@[simps]
+def Pi.evalNonUnitalRingHom (f : I → Type v) [∀ i, NonUnitalNonAssocSemiring (f i)] (i : I) :
+    (∀ i, f i) →ₙ+* f i :=
+  { Pi.evalMulHom f i, Pi.evalAddMonoidHom f i with }
+#align pi.eval_non_unital_ring_hom Pi.evalNonUnitalRingHom
+
+/-- `function.const` as a `NonUnitalRingHom`. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Pi.lean:141]:
```suggestion
/-- `Function.const` as a `NonUnitalRingHom`. -/
```

File: Mathlib/Algebra/Ring/Pi.lean [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+end Pi
+
+section NonUnitalRingHom
+
+universe u v
+
+variable {I : Type u}
+
+/-- Evaluation of functions into an indexed collection of non-unital rings at a point is a
+non-unital ring homomorphism. This is `function.eval` as a `NonUnitalRingHom`. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Pi.lean:134]:
```suggestion
non-unital ring homomorphism. This is `Function.eval` as a `NonUnitalRingHom`. -/
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**